pipeline/artifactory/artifactory.go

```diff
@@ -0,0 +1,429 @@
+// Package artifactory provides a Pipe that push to artifactory
+package artifactory
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"html/template"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/goreleaser/goreleaser/config"
+	"github.com/goreleaser/goreleaser/context"
+	"github.com/goreleaser/goreleaser/internal/buildtarget"
+	"github.com/goreleaser/goreleaser/pipeline"
+
+	"github.com/apex/log"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+)
+
+// artifactoryResponse reflects the response after an upload request
+// to Artifactory.
+type artifactoryResponse struct {
+	Repo              string               `json:"repo,omitempty"`
+	Path              string               `json:"path,omitempty"`
+	Created           string               `json:"created,omitempty"`
+	CreatedBy         string               `json:"createdBy,omitempty"`
+	DownloadURI       string               `json:"downloadUri,omitempty"`
+	MimeType          string               `json:"mimeType,omitempty"`
+	Size              string               `json:"size,omitempty"`
+	Checksums         artifactoryChecksums `json:"checksums,omitempty"`
+	OriginalChecksums artifactoryChecksums `json:"originalChecksums,omitempty"`
+	URI               string               `json:"uri,omitempty"`
+}
+
+// artifactoryChecksums reflects the checksums generated by
+// Artifactory
+type artifactoryChecksums struct {
+	SHA1   string `json:"sha1,omitempty"`
+	MD5    string `json:"md5,omitempty"`
+	SHA256 string `json:"sha256,omitempty"`
+}
+
+const (
+	modeBinary  = "binary"
+	modeArchive = "archive"
+)
+
+// Pipe for Artifactory
+type Pipe struct{}
+
+// String returns the description of the pipe
+func (Pipe) String() string {
+	return "releasing to Artifactory"
+}
+
+// Default sets the pipe defaults
+func (Pipe) Default(ctx *context.Context) error {
+	if l := len(ctx.Config.Artifactories); l == 0 {
+		return nil
+	}
+
+	// Check if a mode was set
+	for i := range ctx.Config.Artifactories {
+		if ctx.Config.Artifactories[i].Mode == "" {
+			ctx.Config.Artifactories[i].Mode = "archive"
+		}
+	}
+
+	return nil
+}
+
+// Run the pipe
+//
+// Docs: https://www.jfrog.com/confluence/display/RTF/Artifactory+REST+API#ArtifactoryRESTAPI-Example-DeployinganArtifact
+func (Pipe) Run(ctx *context.Context) error {
+	if l := len(ctx.Config.Artifactories); l == 0 {
```

Comment:
could be simplified here as welll

Reply:
Again the same as above :)

Reply:
Done in https://github.com/goreleaser/goreleaser/pull/429/commits/4e4529bbf8009ab762b0009766987d57ad5cc1ff